### ISSUE TITLE
[Bug]: Querynode terminated: internal error expr processed rows not equal expect batch size and panic related to knowhere

### ISSUE THREAD

**Issue by ThreadDao** [milvus-io/milvus#29576]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: master-20231227-e53a1aaf
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka):  pulsar 
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus 2.4.0rc4
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

1. deploy config
```
  config:
    dataCoord:
      enableActiveStandby: true
      segment:
        enableLevelZero: true
    dataNode:
      dataSync:
        maxParallelSyncTaskNum: 32
    indexCoord:
      enableActiveStandby: true
    log:
      level: debug
    mixCoord:
      enableActiveStandby: true
    queryCoord:
      enableActiveStandby: true
    rootCoord:
      enableActiveStandby: true
```
2.  create collection -> hnsw index -> insert 3m-128d data into 3 partitions, 1million in each partition -> flush -> load
3. concurrent search + delete + flush + scene_insert_partition
   - delete expr: id > 3000000
   - scene_insert_partition: each locust user insert 200 entities into a random partition (total 10 partitions), and the inserted pk value starts to increase from 3000000
![image](https://github.com/milvus-io/milvus/assets/27288593/feea1088-e7f0-4a0d-b119-87c266357b29)

4. querynode restarted many times
 -  `level-zero-multi-par-1-milvus-querynode-84dd85f888-wdsdz`
 [loki-qn-wdsdz-panic-logs](https://grafana-4am.zilliz.cc/explore?orgId=1&left=%7B%22datasource%22:%22Loki%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bcluster%3D%5C%224am%5C%22,namespace%3D%5C%22qa-milvus%5C%22,pod%3D~%5C%22level-zero-multi-par-1-milvus-querynode-84dd85f888-wdsdz%5C%22%7D%20%7C%3D%20%5C%22Milvus%5C%22%22,%22queryType%22:%22range%22,%22hide%22:false%7D,%7B%22refId%22:%22B%22,%22expr%22:%22%7Bcluster%3D%5C%224am%5C%22,namespace%3D%5C%22qa-milvus%5C%22,pod%3D~%5C%22level-zero-multi-par-1-milvus-querynode-84dd85f888-wdsdz%5C%22%7D%20%7C%3D%20%5C%22panic%5C%22%22,%22queryType%22:%22range%22,%22hide%22:false%7D%5D,%22range%22:%7B%22from%22:%221703690280000%22,%22to%22:%221703712600000%22%7D%7D)
![6tSxgoKX1p](https://github.com/milvus-io/milvus/assets/27288593/0965317c-1a58-4561-a636-8a27da88087d)
 
 - `level-zero-multi-par-1-milvus-querynode-84dd85f888-wm8g9`
   [loki-qn-wm8g9-panic-logs](https://grafana-4am.zilliz.cc/explore?orgId=1&left=%7B%22datasource%22:%22Loki%22,%22queries%22:%5B%7B%22refId%22:%22C%22,%22expr%22:%22%7Bcluster%3D%5C%224am%5C%22,namespace%3D%5C%22qa-milvus%5C%22,pod%3D~%5C%22level-zero-multi-par-1-milvus-querynode-84dd85f888-wm8g9%5C%22%7D%20%7C%3D%20%5C%22panic%5C%22%22,%22queryType%22:%22range%22,%22hide%22:false%7D,%7B%22refId%22:%22D%22,%22expr%22:%22%7Bcluster%3D%5C%224am%5C%22,namespace%3D%5C%22qa-milvus%5C%22,pod%3D~%5C%22level-zero-multi-par-1-milvus-querynode-84dd85f888-wm8g9%5C%22%7D%20%7C%3D%20%5C%22Milvus%5C%22%22,%22queryType%22:%22range%22,%22hide%22:false%7D%5D,%22range%22:%7B%22from%22:%221703690280000%22,%22to%22:%221703712600000%22%7D%7D)

![KanLDSuShQ](https://github.com/milvus-io/milvus/assets/27288593/7701cbb7-1bbc-4209-83b8-759cc7578d86)

### Steps To Reproduce

```markdown
argo: [level-zero-stable-3-1356363395](https://argo-workflows.zilliz.cc/archived-workflows/qa/bd088fa1-adb1-42f3-8eda-3ddb9238ec88?nodeId=level-zero-stable-3-1356363395)
```

### Milvus Log

pods:
```
level-zero-multi-par-1-etcd-0                                     0/1     Pending            0                 22h     <none>          <none>       <none>           <none>
level-zero-multi-par-1-etcd-1                                     1/1     Running            0                 23h     10.104.30.193   4am-node38   <none>           <none>
level-zero-multi-par-1-etcd-2                                     1/1     Running            0                 23h     10.104.34.112   4am-node37   <none>           <none>
level-zero-multi-par-1-milvus-datanode-dc74898f7-hcbmg            1/1     Running            0                 23h     10.104.34.130   4am-node37   <none>           <none>
level-zero-multi-par-1-milvus-datanode-dc74898f7-hsg6m            1/1     Running            0                 23h     10.104.30.210   4am-node38   <none>           <none>
level-zero-multi-par-1-milvus-indexnode-67f86fb574-6lf9p          1/1     Running            0                 23h     10.104.30.211   4am-node38   <none>           <none>
level-zero-multi-par-1-milvus-indexnode-67f86fb574-9gxgw          1/1     Running            0                 23h     10.104.17.197   4am-node23   <none>           <none>
level-zero-multi-par-1-milvus-mixcoord-5b4c869696-j8gs8           1/1     Running            0                 23h     10.104.30.209   4am-node38   <none>           <none>
level-zero-multi-par-1-milvus-proxy-577959c5d4-t97ts              1/1     Running            0                 23h     10.104.17.195   4am-node23   <none>           <none>
level-zero-multi-par-1-milvus-querynode-84dd85f888-wdsdz          1/1     Running            9 (17h ago)       23h     10.104.34.131   4am-node37   <none>           <none>
level-zero-multi-par-1-milvus-querynode-84dd85f888-wm8g9          1/1     Running            7 (17h ago)       23h     10.104.15.200   4am-node20   <none>           <none>
level-zero-multi-par-1-minio-0                                    1/1     Running            0                 23h     10.104.27.167   4am-node31   <none>           <none>
level-zero-multi-par-1-minio-1                                    1/1     Running            0                 23h     10.104.30.191   4am-node38   <none>           <none>
level-zero-multi-par-1-minio-2                                    0/1     Pending            0                 22h     <none>          <none>       <none>           <none>
level-zero-multi-par-1-minio-3                                    1/1     Running            0                 23h     10.104.34.110   4am-node37   <none>           <none>
level-zero-multi-par-1-pulsar-bookie-0                            1/1     Running            0                 23h     10.104.34.111   4am-node37   <none>           <none>
level-zero-multi-par-1-pulsar-bookie-1                            1/1     Running            0                 23h     10.104.30.196   4am-node38   <none>           <none>
level-zero-multi-par-1-pulsar-bookie-2                            0/1     Pending            0                 22h     <none>          <none>       <none>           <none>
level-zero-multi-par-1-pulsar-bookie-init-c644n                   0/1     Completed          0                 23h     10.104.5.153    4am-node12   <none>           <none>
level-zero-multi-par-1-pulsar-broker-0                            1/1     Running            1 (18h ago)       23h     10.104.27.164   4am-node31   <none>           <none>
level-zero-multi-par-1-pulsar-proxy-0                             1/1     Running            2 (16h ago)       23h     10.104.12.206   4am-node17   <none>           <none>
level-zero-multi-par-1-pulsar-pulsar-init-8tfz4                   0/1     Completed          0                 23h     10.104.30.185   4am-node38   <none>           <none>
level-zero-multi-par-1-pulsar-recovery-0                          1/1     Running            3 (18h ago)       23h     10.104.5.154    4am-node12   <none>           <none>
level-zero-multi-par-1-pulsar-zookeeper-0                         0/1     Pending            0                 22h     <none>          <none>       <none>           <none>
level-zero-multi-par-1-pulsar-zookeeper-1                         1/1     Running            0                 23h     10.104.15.189   4am-node20   <none>           <none>
level-zero-multi-par-1-pulsar-zookeeper-2                         1/1     Running            0                 23h     10.104.34.128   4am-node37   <none>           <none>
```

**Comment by zhagnlu** [comment]
maybe same as: https://github.com/milvus-io/milvus/issues/29319, fix pr: https://github.com/milvus-io/milvus/pull/29541

**Comment by chyezh** [comment]
> maybe same as: #29319, fix pr: #29541

/assign @ThreadDao
please help to verify the fixup.
